go/vt/vtgate/planbuilder/operators/update.go

```diff
@@ -200,15 +214,44 @@ func createUpdateWithInputOp(ctx *plancontext.PlanningContext, upd *sqlparser.Up
 	return op
 }
 
-func createUpdateOpWithTarget(ctx *plancontext.PlanningContext, target semantics.TableSet, updStmt *sqlparser.Update) dmlOp {
-	var updExprs sqlparser.UpdateExprs
-	for _, ue := range updStmt.Exprs {
-		if ctx.SemTable.DirectDeps(ue.Name) == target {
-			updExprs = append(updExprs, ue)
+func prepareUpdateExpressionList(ctx *plancontext.PlanningContext, upd *sqlparser.Update) map[semantics.TableSet]updList {
+	// Any update expression requiring column value from any other table is rewritten to take it as bindvar column.
+	// E.g. UPDATE t1 join t2 on t1.col = t2.col SET t1.col = t2.col + 1 where t2.col = 10;
+	// SET t1.col = t2.col + 1 -> SET t1.col = :t2_col + 1 (t2_col is the bindvar column which will be provided from the input)
+	ueMap := make(map[semantics.TableSet]updList)
+	var dependentCols updList
+	for _, ue := range upd.Exprs {
+		target := ctx.SemTable.DirectDeps(ue.Name)
+		exprDeps := ctx.SemTable.RecursiveDeps(ue.Expr)
+		jc := breakExpressionInLHSandRHS(ctx, ue.Expr, exprDeps.Remove(target))
+		updCol := updColumn{ue.Name, jc}
+		ueMap[target] = append(ueMap[target], updCol)
+		dependentCols = append(dependentCols, updCol)
```

Comment:
It looks like `dependentCols` is actually the same as `ueMap[target]`. We can just use that instead of creating this variable.